### DISCUSSION TITLE
The function size exceeds the allowable limit

### DISCUSSION THREAD

**Discussion by akoskm** [supabase/supabase#11651] [Questions]
Hi!

I'm having trouble deploying a simple serverless function that reads from one of my supabase tables, and creates a simple pdf file:

```ts
// Follow this setup guide to integrate the Deno language server with your editor:
// https://deno.land/manual/getting_started/setup_your_environment
// This enables autocomplete, go to definition, etc.

import { serve } from "https://deno.land/std@0.131.0/http/server.ts";
import { createClient } from 'https://esm.sh/@supabase/supabase-js@2';
import html_to_pdf from 'https://esm.sh/html-pdf-node@1.0.8/?target=deno&no-check';

console.log("Hello from Functions!")

serve(async (req: Request) => {
  try {
    const { id } = await req.json()

    // Create a Supabase client with the Auth context of the logged in user.
    const supabaseClient = createClient(
      // Supabase API URL - env var exported by default.
      Deno.env.get('SUPABASE_URL') ?? '',
      // Supabase API ANON KEY - env var exported by default.
      Deno.env.get('SUPABASE_SERVICE_ROLE_KEY') ?? '',
    )

    const { data, error } = await supabaseClient.from("cv").select("*").eq("id", id);
    if (error) throw error;

    if (data.length < 1) return new Response('Not Found', { status: 404});

    const [employee] = data
    const name = `${employee.first_name} ${employee.last_name}`;

    const options = { format: 'A4' };

    const file = { content: `<h1>Welcome ${name}</h1>` };

    html_to_pdf.generatePdf(file, options).then((pdfBuffer: Buffer) => {
      console.log("PDF Buffer:-", pdfBuffer);
    });

    const uploadName = `${employee.first_name}-${employee.last_name}-${new Date()}`;
    const { data: upload } = await supabaseClient.storage
      .from('pdfs')
      .upload(uploadName, '', {
        contentType: 'application/pdf',
        cacheControl: '3600',
        upsert: false,
      });

    return new Response(
      JSON.stringify(employee),
      { headers: { "Content-Type": "application/json" } },
    )
  } catch (error) {
    return new Response(
      JSON.stringify(error),
      { headers: { "Content-Type": "application/json" }, status: 500 },
    )
  }
})
```

This happens when I'm running `supabase functions deploy`:

```
% supabase functions deploy hello-world --project-ref jkbhsduisiudgnsklgkdg
Bundling hello-world
Error: Failed to create a new Function on the Supabase project: {"message":"The function size exceeds the allowable limit"}
Try rerunning the command with --debug to troubleshoot the error.
```

I suspect it's related to the `import html_to_pdf from 'https://esm.sh/html-pdf-node@1.0.8/?target=deno&no-check';` import, because without that file and the PDF generation, the function deploys.

According to https://www.npmjs.com/package/html-pdf-node the whole thing is 56.9 kB unpacked.

What size constraints do I have on the Free Supabase plan for Edge Functions?

**Comment by jonathanstanley** [comment]
if you scroll down on the [pricing page](https://supabase.com/pricing) to edge functions, there is a 2mb limit on free plan. You can also see the [limits here.](https://github.com/supabase/supabase/blob/132022225fe4123fdb96f3a5ad6714f87eaf61fd/apps/www/data/Pricing.json#L219-L223)

**Reply by akoskm** [reply]
Thank you! I suspected there is a size limit, but I'm not sure how I'm hitting it when the script I'm using is only 56.9 kB.

Can you explain how the size is calculated?

**Reply by jonathanstanley** [reply]
when this happened to me, i remember reading that it includes the size of the entire dependency tree... not just your own code. i think it may give clarification in the logs

**Reply by akoskm** [reply]
What's the best way to get this metric for cloud functions?

**Reply by cmtoomey** [reply]
I'm having this same issue - but when deploying via [GH Action](https://github.com/supabase/supabase/blob/master/examples/edge-functions/supabase/functions/github-action-deploy/README.md).

The function deploys just fine from the CLI, but errors out when using an Action runner. Running it locally shows `(script size: 989.6kB)` - which is well under the 2mb limit.

**Comment by ilbertt** [comment]
Hi @akoskm, how did you manage to make `html-pdf-node` work in Supabase's Deno runtime? I'm getting this error, because the runtime [doesn't allow filesystem access](https://supabase.com/docs/guides/functions/debugging#limitations):
```
TypeError: Deno.readFileSync is not a function
    at Object.readFileSync (https://deno.land/std@0.177.1/node/_fs/_fs_readFile.ts:35:21)
    at https://esm.sh/v133/vm2@3.9.19/deno/vm2.mjs:35:3377
    at https://esm.sh/v133/vm2@3.9.19/deno/vm2.mjs:27:458
    at https://esm.sh/v133/vm2@3.9.19/deno/vm2.mjs:42:3481
    at https://esm.sh/v133/vm2@3.9.19/deno/vm2.mjs:27:458
    at https://esm.sh/v133/vm2@3.9.19/deno/vm2.mjs:42:3845
    at https://esm.sh/v133/vm2@3.9.19/deno/vm2.mjs:27:458
    at https://esm.sh/v133/vm2@3.9.19/deno/vm2.mjs:42:4022

failed to send request to user worker: connection error: Connection reset by peer (os error 104)
InvalidWorkerResponse: user worker failed to respond
    at async Promise.all (index 1)
    at async UserWorker.fetch (ext:sb_user_workers/user_workers.js:64:19)
    at async Server.<anonymous> (file:///home/deno/main/index.ts:128:12)
    at async Server.#respond (https://deno.land/std@0.182.0/http/server.ts:220:18) {
  name: "InvalidWorkerResponse"
}
```

How did you solve this problem?

**Reply by akoskm** [reply]
@Luca8991 I'm not on the project anymore, but I believe the team moved the entire app to DigitalOcean.
